### DISCUSSION TITLE
Handling Shutdown of NPM app.

### DISCUSSION THREAD

**Discussion by StefanJonssonInExchange** [microsoft/aspire#6611] [General]
Hi!

I'm using Aspire for a project that contains a C# API. There is also a Vue frontend that is a separate repo.

In production the Vue frontend talks to this API in question. I wanted to integrate the frontend to the already existing aspire project (for local environment purposes). Here is what I did

- Installed the package Aspire.Hosting.NodeJs
- Add the following code to Program.cs in AppHost.

```
builder.AddNpmApp("vueFrontend", "..\\..\\..\\..\\repo-name\\src\\Web")
.WithHttpEndpoint(targetPort: 8088, port: 8088, isProxied: false)
.WithEnvironment(ctx =>
{
  var ep = api.Resource.GetEndpoint("http");
  ctx.EnvironmentVariables["backendBaseUrl"] = ReferenceExpression.Create(
      $"{ep.Property(EndpointProperty.Scheme)}://{ep.Property(EndpointProperty.IPV4Host)}:{ep.Property(EndpointProperty.Port)}");
});
```

This works pretty good and was easy to accomplish. However, when I stop debugging/close the dashboard. My API, and other services like rabbitmq, cosmos emulator etc are closed successfully. However the frontend is still running. And I can see in my Task manager that Node.js is still running, (and starting another instance on each debugging session)

The problem with this is that sometimes (very inconsistent) the entire dashboard cannot launch unless i close the node.js instances in task manager, with this exception:

`Failed to bind to address http://127.0.0.1:15282: address already in use.`

However, that url is to the Aspire dashboard and not the frontend application, so I'm abit confused.

Anyways. If I manually always kill node.js instances in task manager before restarting my debugging session I cannot reproduce the problem. And I can also not reproduce the problem if I comment out the vue frontend in my AppHost. So I am pretty sure the problem is that Aspire is not shutting down the frontend application.

When we debug the frontend without aspire, we start it with npm run start, and we close it by just pressing ctrl+c in the terminal.

Screenshot of package.json in frontend repo.
![image](https://github.com/user-attachments/assets/1e801021-96b1-4990-98bc-7340381aee91)

So my question is; Is Aspire not able to shutdown NPM apps properly, or am I doing something wrong perhaps? Would appreciate some advice here. Thanks In advance.

**Comment by davidfowl** [comment]
What version of aspire are you using?

**Comment by StefanJonssonInExchange** [comment]
All Aspire nugets are 9.0.0-rc.1.24511.1 

and AppHost csproj has `<Sdk Name="Aspire.AppHost.Sdk" Version="9.0.0-rc.1.24511.1" />`

**Reply by davidfowl** [reply]
The RC1 has an issue with nodejs shutdown, see https://github.com/dotnet/aspire/issues/6437

**Reply by StefanJonssonInExchange** [reply]
Yep, upgrading solved the problem. 👍 Thanks for quick response!
